src/router_node.c

```diff
@@ -1829,11 +1838,16 @@ static void CORE_delivery_update(void *context, qdr_delivery_t *dlv, uint64_t di
         // If the delivery is still arriving, don't push out the disposition change yet.
         //
         if (qd_message_receive_complete(msg)) {
-            pn_delivery_update(pnd, disp);
+            if (disp != pn_delivery_local_state(pnd)) {
+                pn_delivery_update(pnd, disp);
+                if (qdr_delivery_disposition(dlv) != disp)
+                    qdr_delivery_set_disposition(dlv, disp);
+            }
         } else {
             // just update the local disposition for now - AMQP_rx_handler will
             // write this to proton once the message is fully received.
-            qdr_delivery_set_disposition(dlv, disp);
+            if (qdr_delivery_disposition(dlv) != disp)
```

Comment:
Same here - this should not be necessary.  You can replace these with an assert(qdr_delivery_disposition(dlv) == disp) to ensure this routine isn't called inappropriately just to be safe.